experiment/src/org/labkey/experiment/api/SampleTypeUpdateServiceDI.java

```diff
@@ -856,10 +859,62 @@ else if (!container.isProject() && container.getProject() != null)
         return ContainerFilter.current(container);
     }
 
-    private Map<Integer, Map<String, Object>> getMaterialMapsWithInput(Map<Integer, Map<String, Object>> keys, User user, Container container, boolean checkCrossFolderData, boolean verifyExisting, boolean skipDetails)
+    public record ExistingRowSelect(TableInfo tableInfo, Set<String> columns, boolean includeParent) {}
+
+    public @NotNull ExistingRowSelect getExistingRowSelect(@Nullable Set<String> dataColumns)
+    {
+        if (!(getQueryTable() instanceof UpdateableTableInfo updatable) || dataColumns == null)
+            return new ExistingRowSelect(getQueryTable(), ALL_COLUMNS, true);
+
+        CaseInsensitiveHashMap<String> remap = updatable.remapSchemaColumns();
+        if (null == remap)
+            remap = CaseInsensitiveHashMap.of();
+
+        Set<String> includedColumns = new CaseInsensitiveHashSet("name", "lsid", "rowid");
+        for (ColumnInfo column : getQueryTable().getColumns())
+        {
+            if (dataColumns.contains(column.getColumnName()))
+                includedColumns.add(column.getColumnName());
+            else if (dataColumns.contains(remap.get(column.getColumnName())))
+                includedColumns.add(remap.get(column.getColumnName()));
+        }
+
+        boolean isAllFromMaterialTable = true;
+        Set<String> materialColNames = new CaseInsensitiveHashSet(Stream.of(ExpMaterialTable.Column.values())
+                .map(Enum::name)
+                .collect(Collectors.toSet()));
+        for (String column : includedColumns)
+            isAllFromMaterialTable = isAllFromMaterialTable && materialColNames.contains(column);
+        TableInfo selectTable = isAllFromMaterialTable ? ExperimentService.get().getTinfoMaterial() : getQueryTable();
+
+        boolean hasParentInput = false;
+        if (_sampleType != null)
+        {
+            try
+            {
+                Map<String, String> importAliases = _sampleType.getImportAliasMap();
+                for (String col : dataColumns)
+                {
+                    if (!hasParentInput && ExperimentService.isInputOutputColumn(col) || equalsIgnoreCase("parent",col) || (importAliases != null && importAliases.containsKey(col)))
```

Comment:
Since `importAliases` is tagged as `@NotNull` do we need the null check?
```suggestion
                    if (!hasParentInput && ExperimentService.isInputOutputColumn(col) || equalsIgnoreCase("parent",col) || importAliases.containsKey(col))
```